File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/collections/CollectionSchema.java
package in.org.quicko.silo.client.collections;

import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.entries.EntryReader;
import in.org.quicko.silo.client.scope.DeleteOptions;
import in.org.quicko.silo.client.scope.ScopeReference;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.TransportRequest;

/**
 * One collection's schema, bundled on every read.
 *
 * <p>Deleting the schema deletes the collection: that is the only path the
 * server exposes for it, which is why {@code delete} lives here rather than on
 * the handle.
 */
public final class CollectionSchema {
  private final ScopeReference scope;
  private final String name;
  private final EntryReader reader;

  public CollectionSchema(ScopeReference scope, String name) {
    this.scope = scope;
    this.name = name;
    this.reader = scope.cache().wrap(new EntryReader(scope, name));
  }

  public CollectionDefinition get() {
    return get(RequestOptions.none());
  }

  public CollectionDefinition get(RequestOptions options) {
    if (options == null) return get(RequestOptions.none());
    options.throwIfCancelled("GET", path());
    String body = reader.getSchema(options);
    options.throwIfCancelled("GET", path());
    return CollectionDefinition.fromWire(scope.transport().codec().tree(body));
  }

  public CollectionDefinition put(JsonSchema schema) {
    return put(schema, RequestOptions.none());
  }

  public CollectionDefinition put(JsonSchema schema, RequestOptions options) {
    return CollectionDefinition.fromWire(scope.transport().json(
        TransportRequest.put(path()).body(schema.document()).options(options).build()));
  }

  /** Deletes the schema, and the collection with it. */
  public void delete() {
    delete(DeleteOptions.none());
  }

  public void delete(DeleteOptions options) {
    scope.transport().empty(
        TransportRequest.delete(path())
            .query("force", options.force() ? true : null)
            .options(options.request())
            .build());
  }

  private String path() {
    return ApiPath.collectionSchema(scope.project(), scope.environment(), name);
  }
}

File: packages/silo-client-dart/lib/src/collections/collection_schema.dart
import '../request_options.dart';
import '../scope/delete_options.dart';
import '../scope/scope_reference.dart';
import '../transport/api_path.dart';
import '../transport/transport_request.dart';
import 'collection_definition.dart';
import 'json_schema.dart';

/// One collection's schema. Deleting it deletes the collection.
final class CollectionSchema {
  const CollectionSchema(this._scope, this._name);

  final ScopeReference _scope;
  final String _name;

  Future<CollectionDefinition> get([RequestOptions options = const RequestOptions()]) async =>
      CollectionDefinition.fromJson(await _scope.transport.json(TransportRequest('GET', _path, options: options)));

  Future<CollectionDefinition> put(JsonSchema schema, [RequestOptions options = const RequestOptions()]) async =>
      CollectionDefinition.fromJson(
        await _scope.transport.json(TransportRequest('PUT', _path, body: schema, options: options)),
      );

  Future<void> delete([DeleteOptions options = const DeleteOptions()]) => _scope.transport.empty(
    TransportRequest(
      'DELETE',
      _path,
      query: {if (options.force) 'force': true},
      options: options,
      evicts: ApiPath.entries(_scope.project, _scope.environment, _name),
    ),
  );

  String get _path => ApiPath.collectionSchema(_scope.project, _scope.environment, _name);
}
